crates/bevy_ui/src/widget/text.rs

```diff
@@ -1,33 +1,135 @@
 use bevy_render::texture::Image;
 use bevy_sprite::TextureAtlas;
 use bevy_text::{
-    Font, FontAtlasSet, FontAtlasWarning, Text, TextError, TextLayoutInfo, TextPipeline,
-    TextSettings, YAxisOrientation,
+    Font, FontAtlasSet, FontAtlasWarning, Text, TextError, TextLayoutInfo, TextMeasureInfo,
+    TextPipeline, TextSettings, YAxisOrientation,
 };
 use bevy_window::{PrimaryWindow, Window};
+use taffy::style::AvailableSpace;
 
 fn scale_value(value: f32, factor: f64) -> f32 {
     (value as f64 * factor) as f32
 }
 
-/// Defines how `min_size`, `size`, and `max_size` affects the bounds of a text
-/// block.
-pub fn text_constraint(min_size: Val, size: Val, max_size: Val, scale_factor: f64) -> f32 {
-    // Needs support for percentages
-    match (min_size, size, max_size) {
-        (_, _, Val::Px(max)) => scale_value(max, scale_factor),
-        (Val::Px(min), _, _) => scale_value(min, scale_factor),
-        (Val::Auto, Val::Px(size), Val::Auto) => scale_value(size, scale_factor),
-        _ => f32::MAX,
+#[derive(Clone)]
+pub struct TextMeasure {
+    pub info: TextMeasureInfo,
+}
+
+impl Measure for TextMeasure {
+    fn measure(
+        &self,
+        width: Option<f32>,
+        height: Option<f32>,
+        available_width: AvailableSpace,
+        available_height: AvailableSpace,
+    ) -> Vec2 {
+        let x = width.unwrap_or_else(|| match available_width {
+            AvailableSpace::Definite(x) => x.clamp(
+                self.info.min_width_content_size.x,
+                self.info.max_width_content_size.x,
+            ),
+            AvailableSpace::MinContent => self.info.min_width_content_size.x,
+            AvailableSpace::MaxContent => self.info.max_width_content_size.x,
+        });
+
+        height
+            .map_or_else(
+                || match available_height {
+                    AvailableSpace::Definite(y) => {
+                        let y = y.clamp(
+                            self.info.max_width_content_size.y,
+                            self.info.min_width_content_size.y,
+                        );
+                        self.info.compute_size(Vec2::new(x, y))
+                    }
+                    AvailableSpace::MinContent => Vec2::new(x, self.info.max_width_content_size.y),
+                    AvailableSpace::MaxContent => Vec2::new(x, self.info.min_width_content_size.y),
+                },
+                |y| Vec2::new(x, y),
+            )
+            .ceil()
+    }
+
+    fn dyn_clone(&self) -> Box<dyn Measure> {
+        Box::new(self.clone())
+    }
+}
+
+/// Creates a `Measure` for text nodes that allows the UI to determine the appropriate amount of space
+/// to provide for the text given the fonts, the text itself and the constraints of the layout.
+pub fn measure_text_system(
+    mut queued_text: Local<Vec<Entity>>,
+    mut last_scale_factor: Local<f64>,
+    fonts: Res<Assets<Font>>,
+    windows: Query<&Window, With<PrimaryWindow>>,
+    ui_scale: Res<UiScale>,
```

Comment:
Yeah it's very difficult to follow with how scale factor is applied and removed all over the place at the moment.

Reply:
The original behaviour with  `scale_factor` being applied in `upsert_leaf` is a bug. The `MeasureFunc` is used by the layout algorithm to calculate how much space to give nodes containing content. It should only ever be updated on changes to the `CalculatedSize` component made by the systems managing that content, and never because of changes to the layout or scale factor.